File: semester_two/algoDat/own/blatt-07/LinkedListItem.cpp
// Copyright 2017, University of Freiburg,
// Chair of Algorithms and Data Structures.
// Author: Hannah Bast <bast@cs.uni-freiburg.de>.

#include "./LinkedListItem.h"
#include <string>
#include <sstream>

// _____________________________________________________________________________
int LinkedListItem::numObjects = 0;

// _____________________________________________________________________________
LinkedListItem::LinkedListItem(const std::string& value) {
  numObjects++;
  this->value = value;
  this->previousItem = NULL;
  this->nextItem = NULL;
  this->id = numObjects;
}

// _____________________________________________________________________________
std::string LinkedListItem::idString(LinkedListItem* item) {
  std::ostringstream os;
  os << "#" << (item != NULL ? item->id : 0);
  return os.str();
}

// _____________________________________________________________________________
std::string LinkedListItem::toString() {
  std::ostringstream os;
  os << "[this = " << idString(this) << ", prev = " << idString(previousItem)
     << ", next = " << idString(nextItem) << ", value = \"" << value << "\"]";
  return os.str();
}

File: semester_two/algoDat/public/code/vorlesung-07/vorlagen/java/LinkedListItem.java
// Copyright 2017, University of Freiburg,
// Chair of Algorithms and Data Structures.
// Author: Hannah Bast <bast@cs.uni-freiburg.de>,
//         Axel Lehmann <lehmann@cs.uni-freiburg.de>.

/**
 * Class for a single linked list item.
 */
public class LinkedListItem {

  /**
   * Create new item with given value.
   */
  public LinkedListItem(String value) {
    numObjects += 1;
    this.value = value;
    this.previousItem = null;
    this.nextItem = null;
    this.id = numObjects;
  }

  /**
   * The value of this item (any string), can be the same for different items.
   */
  protected String value;

  /**
   * An internal id, for debugging purposes only.
   */
  protected int id;

  /**
   * An internal id, for debugging purposes only.
   */
  static int numObjects = 0;

  /**
   * The id as human-readable string.
   */
  static String id(LinkedListItem item) {
    return item != null ? ("#" + item.id) : "#0";
  }

  /**
   * The item previous to this one, null if first item.
   */
  protected LinkedListItem previousItem;

  /**
   * The item next to this one, null if last item.
   */
  protected LinkedListItem nextItem;

  /**
   * This item as human-readable string.
   */
  @Override
  public String toString() {
    return "[this = " + id(this) + ", prev = " + id(previousItem)
      + ", next = " + id(nextItem) + ", value = \"" + value + "\"]";
  }
}
